components/style/properties/longhand/box.mako.rs

```diff
@@ -851,3 +851,79 @@ ${helpers.single_keyword("-moz-appearance",
                          gecko_ffi_name="mAppearance",
                          gecko_constant_prefix="NS_THEME",
                          products="gecko")}
+
+// Non-standard: https://developer.mozilla.org/en-US/docs/Web/CSS/-moz-binding
+<%helpers:longhand name="-moz-binding" products="gecko">
+    use cssparser::{CssStringWriter, ToCss};
+    use gecko_bindings::ptr::{GeckoArcPrincipal, GeckoArcURI};
+    use std::fmt::{self, Write};
+    use url::Url;
+    use values::computed::ComputedValueAsSpecified;
+
+    #[derive(PartialEq, Clone, Debug, HeapSizeOf)]
+    pub struct UrlExtraData {
+        pub base: GeckoArcURI,
+        pub referrer: GeckoArcURI,
+        pub principal: GeckoArcPrincipal,
+    }
+
+    #[derive(PartialEq, Clone, Debug, HeapSizeOf)]
+    pub enum SpecifiedValue {
+        Url(Url, UrlExtraData),
+        None,
+    }
+
+    impl ComputedValueAsSpecified for SpecifiedValue {}
+
+    impl ToCss for SpecifiedValue {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            match *self {
+                SpecifiedValue::Url(ref url, _) => {
+                    try!(dest.write_str("url(\""));
+                    try!(write!(&mut CssStringWriter::new(dest), "{}", url));
+                    try!(dest.write_str("\")"));
```

Comment:
Can we implement ToCss for Url somewhere, and then just delegate to that here (and in the other places where we need to do this)? Followup patch is fine.